shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Chain.hs

```diff
@@ -0,0 +1,159 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Cardano.Ledger.Chain
+  ( -- | Chain Checks
+    ChainChecksData (..),
+    ChainPredicateFailure (..),
+    pparamsToChainChecksData,
+    chainChecks,
+    -- | Ada Pots
+    AdaPots (..),
+    totalAdaES,
+    totalAdaPotsES,
+  )
+where
+
+import Cardano.Ledger.Coin (Coin (..))
+import Cardano.Ledger.Era (Crypto, Era)
+import Cardano.Ledger.Shelley.Constraints (UsesValue)
+import qualified Cardano.Ledger.Val as Val
+import Cardano.Protocol.TPraos.BHeader
+  ( BHeader,
+    bHeaderSize,
+    bhbody,
+    hBbsize,
+  )
+import Control.Monad (unless)
+import Control.Monad.Except (MonadError, throwError)
+import Data.Foldable (fold)
+import qualified Data.Map.Strict as Map
+import GHC.Generics (Generic)
+import GHC.Records
+import NoThunks.Class (NoThunks (..))
+import Numeric.Natural (Natural)
+import Shelley.Spec.Ledger.LedgerState
+  ( AccountState (..),
+    DPState (..),
+    DState (..),
+    EpochState (..),
+    LedgerState (..),
+    UTxOState (..),
+  )
+import Shelley.Spec.Ledger.PParams (ProtVer (..))
+import Shelley.Spec.Ledger.UTxO (balance)
+
+data ChainChecksData = ChainChecksData
```

Comment:
Maybe call this `ChainChecksPParms` to make it clear that this is a subset of the protocol parameters?